connmgr/banmanager.go

```diff
@@ -0,0 +1,202 @@
+func NewBanManager(cfg *BanManagerConfig) *BanManager {
+	return &BanManager{
+		cfg:       *cfg,
+		peers:     make(map[string]*banMgrPeer, cfg.MaxPeers),
+		banScores: make(map[string]*DynamicBanScore, cfg.MaxPeers),
+	}
+}
+
+// IsPeerWhitelisted checks if the provided peer is whitelisted.
+func (bm *BanManager) IsPeerWhitelisted(p *peer.Peer) bool {
+	host, _, err := net.SplitHostPort(p.Addr())
+	if err != nil {
+		log.Errorf("Unable to split peer '%s' IP: %v", p.Addr(), err)
+		return false
+	}
+
+	ip := net.ParseIP(host)
+	if ip == nil {
+		log.Errorf("Unable to parse IP '%s'", p.Addr())
+		return false
+	}
+
+	for _, ipnet := range bm.cfg.WhiteList {
+		if ipnet.Contains(ip) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// AddPeer adds the provided peer to the ban manager.
+func (bm *BanManager) AddPeer(p *peer.Peer) {
+	bmp := &banMgrPeer{
+		Peer:          p,
+		isWhitelisted: bm.IsPeerWhitelisted(p),
+	}
+
+	bm.mtx.Lock()
+	bm.peers[p.Addr()] = bmp
+	bm.banScores[p.Addr()] = &DynamicBanScore{}
+	bm.mtx.Unlock()
+}
+
+// LookupPeer returns the associated peer of the provided address and its
+// ban score.
+func (bm *BanManager) LookupPeer(addr string) (*peer.Peer, uint32) {
+	bm.mtx.Lock()
+	p, ok := bm.peers[addr]
+	if !ok {
+		bm.mtx.Unlock()
+		return nil, 0
+	}
+	score := bm.banScores[addr]
+	bm.mtx.Unlock()
+
+	return p.Peer, score.Int()
+}
+
+// RemovePeer discards the provided peer from the ban manager.
+func (bm *BanManager) RemovePeer(p *peer.Peer) {
+	bm.mtx.Lock()
+	delete(bm.peers, p.Addr())
+	delete(bm.banScores, p.Addr())
+	bm.mtx.Unlock()
+}
+
+// BanPeer bans the provided peer.
+func (bm *BanManager) BanPeer(p *peer.Peer) {
+	// Return immediately if banning is disabled.
+	if bm.cfg.DisableBanning {
+		return
+	}
+
+	bm.mtx.Lock()
+	bmp, ok := bm.peers[p.Addr()]
+	if !ok {
+		bm.mtx.Unlock()
+		log.Errorf("No peer found with IP: '%s'", p.Addr())
+		return
+	}
+	bm.mtx.Unlock()
+
+	// Return if the peer is whitelisted.
+	if bmp.isWhitelisted {
+		return
+	}
+
+	// Ban and remove the peer.
+	bm.cfg.BanPeer(p)
+	bm.RemovePeer(p)
+}
+
+// AddBanScore increases the persistent and decaying ban scores of the
+// provided peer by the values passed as parameters. If the resulting score
+// exceeds half of the ban threshold, a warning is logged including the reason
+// provided. Further, if the score is above the ban threshold, the peer will
+// be banned.
+func (bm *BanManager) AddBanScore(p *peer.Peer, persistent, transient uint32, reason string) bool {
+	// No warning is logged and no score is calculated if banning is disabled.
+	if bm.cfg.DisableBanning {
+		return false
+	}
+
+	bm.mtx.Lock()
+	bmp, ok := bm.peers[p.Addr()]
+	if !ok {
+		bm.mtx.Unlock()
+		log.Debugf("No peer found with IP: '%s'", p.Addr())
+		return false
+	}
+	bm.mtx.Unlock()
+
+	if bmp.isWhitelisted {
+		log.Debugf("Misbehaving whitelisted peer %s: %s", p, reason)
+		return false
+	}
+
+	banScore, ok := bm.banScores[p.Addr()]
```

Comment:
Naturally this all would just become `banScore := &bmp.banScore` with the changes mentioned elsewhere.